### ISSUE TITLE
Better approximation for darkening layers under coat

### ISSUE THREAD

**Issue by AdrienHerubel** [AcademySoftwareFoundation/OpenPBR#136]
Due to the possibility of total internal reflection at the coat/external medium boundary, layers under the coat should be darker. See https://graphics.cs.yale.edu/sites/default/files/wet.pdf for a detailed model describing the various effect of having a water layer on top of materials. I think that interpenetration of layers is out of scope for OpenPBR surface, but the total internal reflection should be modelled.

In the OpenPBR specification we describe the phenomenon without providing implementation details :
`In the full light transport this observed color is further darkened and saturated due to multiple internal reflections from the inside of the coat, including a considerable amount of total internal reflection, which causes light to strike the underlying material multiple times and undergo more absorption. Also the observed tint color should vary away from coat_color as the incidence angle changes, due to the change in path length in the medium. The presence of a rough coat will increase the apparent roughness of the BSDF lobes of the underlying base. We generally assume that in the ground truth appearance, all these effects are accounted for.`

In Standard Surface there is a dedicated parameter to simulate the darkening of layers under coat using the following formula :
`base_color = pow(base_color, 1.0 + (coat * coat_affect_color)) subsurface_color = pow(subsurface_color, 1.0 + (coat * coat_affect_color)) 
`
This approximation is simple and disabled by default in Standard Surface. 

In OpenPBR we should evaluate it against ground truth, possibly find a better approximation, and decide whether we want to make the effect user controllable.

**Comment by portsmouth** [comment]
For the moment, I think it's fine to use the crude Standard Surface model, since according to the spec an implementation can approximate the effect however it wants (including just ignoring it..).

I'd like to leave the effect on since the spirit of the spec is to do this physically correctly, in which case there should be a darkening/saturation effect. Even with the rough approximation, it will be good to get feedback about this and because (as Indy argued) it probably is generally more useful for the effect to be on than off. 

I have a slight preference for using a lower hard-coded value than 1 though (e.g. 0.5), since the effect is quite strong right now, and that may not be realistic.

As noted in the OpenPBR thread, I think we should for a later release derive a slightly better approximation to get more insight into how big this effect actually is. For example if you restrict to normal incidence, I think the full multi-bounce calculation is pretty tractable.

Whether we put in some control later to modulate the effect, is tricky since we would have to give some concrete physical explanation of it.  But once we have a better understanding of a formula for the effect, we could implement it for example like:

> this `coat_darkening_modulation` control works (when set to 1) by altering the base albedo (`base_color`) according to the following formula (which we derive) which approximately makes the apparent color match the input after interaction with the coat. A value of 0 leaves the base_albedo untouched, and values in between blend between these extremes. 

In essence that is quite similar to the scheme we have for SSS remapping (i.e. set the internal albedo/MFP to approximately match the desired color, given a detailed understanding of the effect).

**Comment by jstone-lucasfilm** [comment]
@AdrienHerubel @portsmouth I'd support the idea of providing an _artistic control_ over the darkening effect of the coat, as I would expect the real-world impact of this darkening to be highly variable, and influenced by the composition and thickness of the coating material.

Just to give one example of an artistic use case that OpenPBR ought to handle, here's a link to a printing company talking about the effect of different lamination processes on the colors of printed paper:
https://www.mag-data.net/en/how-lamination-affects-the-colour-rendering-of-a-printed-sheet

In the short term, would it perhaps make sense to add a control along the lines of the `coat_darkening_modulation` that @portsmouth suggests above, so that artists have some degree of control over the current approximation, which can be extended to a more principled approximation in the future?

**Comment by portsmouth** [comment]
Just to clarify, I think if we did add that control back, we would need to present a specific formula in the spec (for remapping the `base_color` to achieve an approximate match). As otherwise it would not be clear how to implement it. It would make sense then to try to derive that formula first, before we make any alterations to the parameter set.

**Comment by AdrienHerubel** [comment]
The setup for the following test is a grey diffuse sphere inside an emissive sphere. The middle images are the same scene with a white coat of roughness 0, IOR 1.5 and three values of coat_affect_color. The image on the right is the same scene with a geometrical coat, fully transmissive with roughness 0.

| Base only | Base with coat coat_affect_color 0 | Base with coat coat_affect_color 0.5 | Base with coat coat_affect_color 1 | Base with modeled coat |
| --- | --- | --- | --- | --- |
| ![base](https://github.com/AcademySoftwareFoundation/OpenPBR/assets/947159/73534c23-5636-481a-be5d-7d5b5a7c75fc) | ![coated_base_af 5](https://github.com/AcademySoftwareFoundation/OpenPBR/assets/947159/934d2ff9-a0ce-4afb-a1b4-92f132ed8728)  | ![coated_base_af0](https://github.com/AcademySoftwareFoundation/OpenPBR/assets/947159/8e4d01b3-5de6-404d-ba9a-1fcbf631d2c9) | ![coated_base_af1](https://github.com/AcademySoftwareFoundation/OpenPBR/assets/947159/af891d27-2679-44cb-92e3-6e9f6550883b) | ![coated_base_simulated](https://github.com/AcademySoftwareFoundation/OpenPBR/assets/947159/c35c9baa-e349-47af-bff0-c13da52151fa) |

We can see that the simulated coat is darker, and even much darker than the coat_affect_color image.

**Comment by jstone-lucasfilm** [comment]
@AdrienHerubel What are your thoughts on how _universal_ this model for coat darkening should be considered?  Would the effect of the lamination process on paper (e.g. a playing card) show such a strong darkening effect as well, or would we expect the composition, application process, and thickness of the coat to have a strong impact on the degree of coat darkening we would see?

**Comment by portsmouth** [comment]
We assume that the coat is a dielectric layer directly on top of the base. Then the effect described in 2.1 of the paper Adrien linked is the only one that occurs. 

![image](https://github.com/AcademySoftwareFoundation/OpenPBR/assets/2774364/a2542f8b-e3f2-4fe4-9608-0b7a0e7cf95d)

That is the effect that Adrien's tests simulate also, by brute force, i.e. bounces in a thin but finite depth coat.

This is actually independent of the thickness of the coat, if the coat is non-absorbing. (If it absorbs, then it depends on the transmittance, not the depth directly).  Even a coat of very small thickness will do this  (though if the coat depth is so small that wave optics is important, then things will be more complicated -- we can just assume that is not the case).

As noted in the paper, in real materials there is also the possibility that the substrate is *wetted* by the coat. Which we could attempt to model, but currently don't (and other CG models don't either I think, currently). Similarly, if the lamination involves some adhesive, we won't/can't model the effect of that.

**Comment by jstone-lucasfilm** [comment]
The main reason that I bring up the potential _variability_ of the effect is make sure we're giving the artist sufficient freedom to control its visual impact, rather than applying a universal darkening to all base colors under a coat.  If there's room for different coats to have different darkening effects, then to my mind that supports the importance of an artistic control to determine where and to what degree it's applied.  If, on the other hand, we expect a uniform, universal darkening effect, then this artistic control would be less critical.

**Comment by AdrienHerubel** [comment]
> Would the effect of the lamination process on paper

The effect of lamination on the color of paper looks well documented https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9504082/

**Comment by jstone-lucasfilm** [comment]
@portsmouth Agreed that there's a visual impact of lamination on printed colors, and I posted a related article above:
https://www.mag-data.net/en/how-lamination-affects-the-colour-rendering-of-a-printed-sheet

But would you expect the visual impact of lamination to share identical math with the visual impact of water films?  That doesn't match my own visual intuition for these two categories of materials, but I'm open to revisiting my expectations if others here believe that the darkening effect ought to be the same in these two cases, and that artists working on materials would share this intuition.

**Comment by jstone-lucasfilm** [comment]
@portsmouth Agreed that there's a visual impact of lamination on printed colors, but I'd expect the impact of this effect to be very different from that of water films on surfaces.  From our current Slack discussion, it sounds like we're leaning in the direction of expecting coatings to have a highly variable effect, very different from the uniform darkening/squaring in Standard Surface, and let's see how the conversation evolves.

**Comment by portsmouth** [comment]
>it sounds like we're leaning in the direction of expecting coatings to have a highly variable effect

Assuming the mechanism is just light bouncing in the dielectric layer, you would expect the effect to vary according to the coat IOR, roughness and color (i.e. absorption transmittance), in a way which is understood and derivable. (Though not very well modeled by the Standard Surface approximation).

**Comment by jstone-lucasfilm** [comment]
@portsmouth Absolutely, I'm expecting that we'll work through the derivation and come up with a new, physically motivated behavior for base-under-coat effects, which we'll then integrate into an upcoming reference implementation for OpenPBR.  Based on its artistic effectiveness, we'll have the choice to enable the new behavior for all OpenPBR materials or to control it with an artist-facing slider, and I'm open to either approach depending on what artists prefer.

In the meantime, though, my hope is that we can remove the legacy Standard Surface behavior (uniform squaring of base-under-coat colors), since it's artistically unintuitive, and doesn't make a good approximation of the new, physically motivated behavior.

What are your thoughts?

**Comment by portsmouth** [comment]
As noted in the PR:

>I'd like to leave the effect on since the spirit of the spec is to do this physically correctly, in which case there should be at least some darkening/saturation effect. Even with the rough approximation, it will be good to get feedback about this and because it probably is generally more useful for the effect to be on than off.

>I have a slight preference for using a lower hard-coded value than 1 though (e.g. 0.5), since the effect is quite strong right now, and that may not be realistic.

**Comment by jstone-lucasfilm** [comment]
@portsmouth I understand your perspective on this, though I feel that our current workaround for base-under-coat effects reduces the quality of the first impressions that artists and developers will get of the OpenPBR shading model.

I'll leave the associated pull request active for now, and I'm open to additional thoughts from developers and artists on this topic:

https://github.com/AcademySoftwareFoundation/OpenPBR/pull/135

**Comment by jstone-lucasfilm** [comment]
@portsmouth I understand your perspective on this, though I feel that our current workaround for base-under-coat effects reduces the quality of the first impressions that artists and developers will have of the OpenPBR shading model.

I'll leave the associated pull request active for now, and I'm open to additional thoughts from developers and artists on this topic:

https://github.com/AcademySoftwareFoundation/OpenPBR/pull/135

**Comment by jstone-lucasfilm** [comment]
@portsmouth I understand your perspective on this, though I feel that our current workaround for base-under-coat effects reduces the quality of the first impressions that artists and developers will have of the OpenPBR shading model.  To my eye, based on experience with earlier shading models such as Standard Surface and ILM Unified, the simple squaring of base colors under a coat looks like a bug rather than an intentional, physically motivated effect.

I'll leave the associated pull request active for now, and I'm open to additional thoughts from developers and artists on this topic:

https://github.com/AcademySoftwareFoundation/OpenPBR/pull/135

**Comment by jstone-lucasfilm** [comment]
@portsmouth I understand your perspective on this, though I feel that our current workaround for base-under-coat effects reduces the quality of the first impressions that artists and developers will have of the OpenPBR shading model

To my eye, having worked with earlier shading models such as Standard Surface and ILM Unified, simply applying a uniform power to base colors under a coat looks like a bug rather than an intentional, physically motivated effect.

I'll leave the associated pull request active for now, and I'd be interested in additional thoughts from developers and artists on this topic:

https://github.com/AcademySoftwareFoundation/OpenPBR/pull/135

**Comment by anderslanglands** [comment]
FWIW from memory at Weta we had a control (on by default) that automatically compensated so the diffuse albedo remained the same at normal incidence.

**Comment by portsmouth** [comment]
Here is a specific suggestion for how we define the parametrization of the coat darkening.

We say that by default, the physical effect of darkening due to internal reflections in the coat should be accounted for. We give a suggested approximate formula for implementing this, which is simply to darken the base lobe by the constant factor (different per channel in general):

![image](https://github.com/AcademySoftwareFoundation/OpenPBR/assets/2774364/11c76f31-50b1-4d4e-9f95-b5c2f566d310)

(where $\eta$ is IOR ratio, $\rho_d$ is the base albedo, and $E_F$ is the average of the Fresnel reflectance). Though implementations are free to improve on this approximation if they wish.

To modulate this effect, we add a new control $\delta$ = **`coat_darkening`**, which defaults to 1. In this case, the physical darkening effect above occurs, as described. In the case **`coat_darkening`**=0, we say that the base albedo is boosted (in each channel) by just enough to *counteract the darkening* (in the zero absorption case).This is always doable in principle, since as the base albedo $\rightarrow 1$, the total albedo $\rightarrow 1$. Then intermediate **`coat_darkening`** values are defined to linearly interpolate the base albedo between these limits.

To be precise about what we  mean by "counteract the darkening",  we have to define it concretely in terms of (say) the physical albedo of the coated base at normal incidence, $R_c(1)$. As for the glossy-diffuse [case](https://github.com/AcademySoftwareFoundation/OpenPBR/issues/141#issuecomment-1853232888), to be precise we can say that the "undarkened" coat should be such that $R_c(1)$ is as follows (i.e. the albedo-scaling approximation of the undarkened coat):

$$
R_c(1) = \mathrm{lerp}(F_0, 1, \rho_d) = F_0 + \rho_d (1 - F_0)
$$

We can then solve for the boosted base albedo $B \rho_d$ (with boost factor $B$), such that after the coat darkening is applied, this required $R_c(1)$ is obtained.  In the Lambertian approximation, this boost factor is just the reciprocal of the darkening factor above, i.e. $B = 1/\Delta$. (Though if we define it as stated, we don't rely on this approximation). 

Thus we can define the base albedo to be boosted by the following variable factor $B(\delta)$ (where $\delta$ = **`coat_darkening`**):

$$
B(\delta) = \mathrm{lerp}(\frac{1}{\Delta}, 1, \delta)
$$

and $\Delta$ (the coat darkening factor) is implementation dependent, but can be reasonably approximated by the formula above). Note that in the approximation where no darkening is applied (i.e. $\Delta=1$), the boost factor can just be omitted.

**Comment by portsmouth** [comment]
This can be closed now.
